sgx-tools/dcap_linux.go

```diff
@@ -0,0 +1,487 @@
+	}
+
+	akIdentity := &attestationKeyIdentity{}
+	if err = restruct.Unpack(akId, binary.LittleEndian, &akIdentity); err != nil {
+		return nil, err
+	}
+
+	if strings.EqualFold(quoteType, quoteTypeEcdsa) && (akIdentity.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP256 || akIdentity.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP384) {
+		return akId, nil
+	} else if (strings.EqualFold(quoteType, quoteTypeEpidUnlinkable) || strings.EqualFold(quoteType, quoteTypeEpidLinkable)) && akIdentity.AlgorithmId == sgxQuoteLibraryAlgorithmEpid {
+		return akId, nil
+	}
+
+	akIdNum, err := getSupportedAttKeyIDNumRequest()
+	if err != nil {
+		return nil, err
+	}
+	if akIdNum == 0 {
+		return nil, fmt.Errorf("failed to call getSupportedAttKeyIDNumRequest: invalid Attestation Key Identity Number = %d", akIdNum)
+	}
+
+	akIdList, err := getSupportedAttKeyIDsRequest(akIdNum)
+	if err != nil {
+		return nil, err
+	}
+	if akIdList == nil {
+		return nil, fmt.Errorf("failed to call getSupportedAttKeyIDsRequest")
+	}
+
+	var i uint32 = 0
+	for i < akIdNum {
+		akId := akIdList[i*attestationKeyIdentityLength : (i+1)*attestationKeyIdentityLength]
+
+		akIdentity := &attestationKeyIdentity{}
+		if err = restruct.Unpack(akId, binary.LittleEndian, &akIdentity); err != nil {
+			return nil, err
+		}
+
+		if strings.EqualFold(quoteType, quoteTypeEcdsa) && (akIdentity.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP256 || akIdentity.AlgorithmId == sgxQuoteLibraryAlgorithmEcdsaP384) {
+			return akId, nil
+		} else if (strings.EqualFold(quoteType, quoteTypeEpidUnlinkable) || strings.EqualFold(quoteType, quoteTypeEpidLinkable)) && akIdentity.AlgorithmId == sgxQuoteLibraryAlgorithmEpid {
+			return akId, nil
+		}
+
+		i++
+	}
+
+	return nil, nil
+}
+
+func GetQeTargetInfoEx(quoteType string) ([]byte, error) {
+	var ti []byte
+
+	akId, err := getAttestationKeyIdentity(quoteType)
+	if err != nil {
+		return nil, err
+	}
+	if len(akId) != attestationKeyIdentityLength {
+		return nil, fmt.Errorf("len(attKeyId) is not %d but %d", attestationKeyIdentityLength, len(akId))
+	}
+
+	if strings.EqualFold(quoteType, quoteTypeEcdsa) {
+		_, bufSize, err := initQuoteExRequest(akId, false, uint64(0))
+		if err != nil {
+			return nil, err
+		}
+		if bufSize == 0 {
+			return nil, fmt.Errorf("failed to call initQuoteExRequest: invalid Public Key Identity Size: %d\n", bufSize)
+		}
+
+		ti, _, err = initQuoteExRequest(akId, true, bufSize)
+		if err != nil {
+			return nil, err
+		}
+		if len(ti) != intelsgx.TargetinfoLength {
+			return nil, fmt.Errorf("failed to call initQuoteExRequest: invalid length of targetinfo (returned %d, expected %d)", len(ti), intelsgx.TargetinfoLength)
+		}
+		if err = DumpTargetInfo(ti); err != nil {
+			return nil, fmt.Errorf("Invalid target info")
+		}
+	} else if strings.EqualFold(quoteType, quoteTypeEpidUnlinkable) || strings.EqualFold(quoteType, quoteTypeEpidLinkable) {
+		ti, err = intelsgx.GetQeTargetInfo()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Just go out this branch and return the result. There is not necessary to explicitly return from here.